src/core/service.c

```diff
@@ -929,8 +929,11 @@ static int service_load_pid_file(Service *s, bool may_warn) {
         prio = may_warn ? LOG_INFO : LOG_DEBUG;
 
         fd = chase_symlinks(s->pid_file, NULL, CHASE_OPEN|CHASE_SAFE, NULL);
-        if (fd == -EPERM)
-                return log_unit_full(UNIT(s), prio, fd, "Permission denied while opening PID file or unsafe symlink chain: %s", s->pid_file);
+        if (fd == -EPERM) {
+                log_unit_full(UNIT(s), prio, fd, "Permission denied while opening PID file or unsafe symlink chain. Fallback to unsafe logic: %s", s->pid_file);
+                /* The following fallback logic is not safe. It may be dropped in the future release. */
+                fd = chase_symlinks(s->pid_file, NULL, CHASE_OPEN, NULL);
```

Comment:
hmm, this fallback is acceptable, but if we do this, then we should enable the more restrictive in-cgroup checking for the PID still. i.e. we should set a bool here, that we check further down when service_is_suitable_main_pid() returns 0: in that case we should *not* accept the PID file if it points to a PID outside of the cgroup.

Explanation: service_is_suitable_main_pid() returns < 0 if the specified PID is not suitable as main PID for the service. It returns > 0 if it is fully safe (i.e. within the service's cgroup). But it reports == 0 if its "questionnable", i.e. not within the cgroup itself. In that case we currently do a simple fstat() on the PID file and if its owned by root accept it either way. But precisely *that* should be bypassed if the symlink was unsafe: we shouldn't allow questionnable PIDs from PID files in unsafe symlink chains even if the final file is owned by root...